pkg/cloud/cloud.go

```diff
@@ -405,6 +405,9 @@ func (c *cloud) DetachDisk(ctx context.Context, volumeID, nodeID string) error {
 
 	_, err = c.ec2.DetachVolumeWithContext(ctx, request)
 	if err != nil {
+		if isAWSErrorVolumeNotFound(err) {
```

Comment:
Is the `ErrNotFound` returned from `c.getInstance(ctx, nodeID)`enough to cover the issue reported by @jsafrane ? Why do we still need this check?

Also, regarding driver is returning internal error, should we check for `IncorrectState` when detaching an volume that is already detached, and return success? It will get following error in this case:
```
An error occurred (IncorrectState) when calling the DetachVolume operation: Volume 'vol-0c1cd2b52306edf24'is in the 'available' state.
```

Reply:
> That's a good point. Are you sure we get a IncorrectState in this case? Looking at the docs, it seems like we get a InvalidAttachment.NotFound instead.

I tested using awscli as follows:
```
aws ec2 detach-volume --volume-id vol-0d63967ab4999f4b6 --region us-east-1 --instanc
e-id i-03376857fb7981add

An error occurred (IncorrectState) when calling the DetachVolume operation: Volume 'vol-0d63967ab4999f4b6'is in the 'available' state.
```
And similar result with a simple go program:
```
func main() {
	c, _ := cloud.NewCloud("us-east-1")
	err := c.DetachDisk(context.Background(), "vol-0d63967ab4999f4b6", "i-03376857fb7981add")

	fmt.Println(err)
}
```

Although I don't know when `InvalidAttachment.NotFound` could happen, how about we check for both?

Reply:
@jsafrane how is the check done in cloud provider now?

Reply:
> Although I don't know when InvalidAttachment.NotFound could happen, how about we check for both?

I think that makes sense. I changed the code to reflect that.

From the docs:

```
InvalidAttachment.NotFound | Indicates an attempt to detach a volume from an instance to which it is not attached.
```

Perhaps `InvalidAttachment.NotFound` is returned when the volume is attached to another instance?

Reply:
> @jsafrane how is the check done in cloud provider now?

Apparently it doesn't special-case any error; it just returns it:

https://github.com/kubernetes/kubernetes/blob/5bac42bff9bfb9dfe0f2ea40f1c80cac47fc12b2/staging/src/k8s.io/legacy-cloud-providers/aws/aws.go#L2392-L2395

That makes sense because the in-tree plugin doesn't any requirement to be idempotent.

Reply:
> Perhaps InvalidAttachment.NotFound is returned when the volume is attached to another instance?
Yep, just tested and confirmed you are right 👍

Reply:
> > @jsafrane how is the check done in cloud provider now?
> 
> Apparently it doesn't special-case any error; it just returns it:
> 
> https://github.com/kubernetes/kubernetes/blob/5bac42bff9bfb9dfe0f2ea40f1c80cac47fc12b2/staging/src/k8s.io/legacy-cloud-providers/aws/aws.go#L2392-L2395
> 
> That makes sense because the in-tree plugin doesn't any requirement to be idempotent.

I'm thinking how about we move the logic of:

```
if isAWSErrorIncorrectState(err) ||
			isAWSErrorInvalidAttachmentNotFound(err) ||
			isAWSErrorVolumeNotFound(err) {
			return ErrNotFound
		}
```
To controller.go `ControllerUnpublishVolume` ?
This has several advantages that: 1) keep the cloud provider simpler and cleaner 2) align with current cloud provider implementation so that we can unify the two cloud providers (legacy cloud provider and the one used in CSI driver) easier See @jsafrane 's suggestion [here](https://github.com/kubernetes-sigs/aws-ebs-csi-driver/issues/393) 3) keep the CSI specific error handling in CSI controller

Reply:
> I'm thinking how about we move the logic of:
> 
> ```
> if isAWSErrorIncorrectState(err) ||
> 			isAWSErrorInvalidAttachmentNotFound(err) ||
> 			isAWSErrorVolumeNotFound(err) {
> 			return ErrNotFound
> 		}
> ```
> 
> To controller.go `ControllerUnpublishVolume` ?
> This has several advantages that: 1) keep the cloud provider simpler and cleaner 2) align with current cloud provider implementation so that we can unify the two cloud providers (legacy cloud provider and the one used in CSI driver) easier See @jsafrane 's suggestion [here](https://github.com/kubernetes-sigs/aws-ebs-csi-driver/issues/393) 3) keep the CSI specific error handling in CSI controller

A drawback would be that the controller will have to understand AWS-specific error codes.

As an alternative, we could implement a helper function in the cloud package (be it the legacy one or this one) to evaluate whether the error means "OK" in an idempotent context. This method would then be used in the CSI driver only, not in the in-tree plugin.

In any case, whatever the solution we choose, we should implement it equally across all the other methods in the cloud package. Unfortunately I don't have the cycles for that at the moment.

Reply:
k. I can leave with the current change for now. We will reconsider this when unifying legacy cloud provider [here](https://github.com/kubernetes-sigs/aws-ebs-csi-driver/issues/393)